### PR TITLE
feat: support handshake on molecule with features

### LINKED ISSUE #180: Support serialization format molecule
Allow messages encoded by molecule. Library user can choose different serialization format via cargo feature.

### PR DESCRIPTION
**Pull request by driftluo**
- [x] close #180
- [x] wait for [molecule](https://github.com/yangby-cryptape/molecule/tree/develop) release
- [x] wait some small code refactoring

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,4 +2,6 @@
 **/*.rs.bk
 .idea/
 Cargo.lock
-*/target
\ No newline at end of file
+*/target
+*/*/Cargo.lock
```
Comment (doitian) [inline .gitignore]: the line "Cargo.lock" should already cover these rules.

File: .travis.yml [hunk 2]
```diff
@@ -9,23 +9,34 @@ script:
...
         - cargo fmt --all -- --check
-    - stage: Check
+    - stage: Clippy Check
       name: Clippy
       script:
-        - RUSTFLAGS='-F warnings' cargo clippy --all --tests
-    - stage: Test
+        - make clippy
+    - stage: Test Molc
+      name: Unitest
```
Comment (doitian) [inline .travis.yml]:
```suggestion
      name: Unitest (molc)
```

File: .travis.yml [hunk 3]
```diff
@@ -9,23 +9,34 @@ script:
   - cargo build --examples --all
 stages:
```
Comment (doitian) [inline .travis.yml]: Jobs in the same stage can run in parallel. The old setup is more reasonable, using just two stages, Check and Test.

File: .travis.yml [hunk 4]
```diff
@@ -9,23 +9,34 @@ script:
...
       script:
-        - RUSTFLAGS='-F warnings' cargo clippy --all --tests
-    - stage: Test
+        - make clippy
+    - stage: Test Molc
+      name: Unitest
+      script:
+        - RUSTFLAGS='-F warnings' RUST_BACKTRACE=full cargo test --all --features molc
+    - stage: Test Molc
+      name: Bench
```
Comment (doitian) [inline .travis.yml]:
```suggestion
      name: Bench (molc)
```

File: .travis.yml [hunk 5]
```diff
@@ -9,23 +9,34 @@ script:
...
+      script:
+        - cd bench
+        - cargo run --release --features molc
+    - stage: Test Flatc
       name: Unitest
```
Comment (doitian) [inline .travis.yml]:
```suggestion
      name: Unitest (flatc)
```

File: .travis.yml [hunk 6]
```diff
@@ -9,23 +9,34 @@ script:
...
       script:
-        - RUSTFLAGS='-F warnings' RUST_BACKTRACE=full cargo test --all
-    - stage: Test
+        - RUSTFLAGS='-F warnings' RUST_BACKTRACE=full cargo test --all --features flatc
+    - stage: Test Flatc
       name: Bench
```
Comment (doitian) [inline .travis.yml]:
```suggestion
      name: Bench (flatc)
```

File: Cargo.toml [hunk 7]
```diff
@@ -36,20 +37,25 @@ features = ["minwindef", "ws2def", "winerror"]
...
 systemstat = "0.1.3"
-ping = { path = "protocols/ping", package = "tentacle-ping" }
-identify = { path = "protocols/identify", package = "tentacle-identify" }
-discovery = { path = "protocols/discovery", package="tentacle-discovery" }
 
 [target.'cfg(unix)'.dev-dependencies]
 nix = "0.13.0"
 
+[features]
+default = []
```
Comment (doitian) [inline Cargo.toml]: Why not choose one as the default?

Comment (driftluo) [reply Cargo.toml]: I think we should have no default tendency, more is user choice.

and with default, dependence must write by:
```toml
p2p = { path = "../..", version = "0.2.0", package = "tentacle", default-features = false, features = ["molc"] }
```

but if there is no default choose:
```
p2p = { path = "../..", version = "0.2.0", package = "tentacle",  features = ["molc"] }
```

Comment (driftluo) [reply Cargo.toml]: I think we should have no default tendency, more is user choice.

and with default, dependence must write by:
```toml
p2p = { path = "../..", version = "0.2.0", package = "tentacle", default-features = false, features = ["molc"] }
```

but if there is no default choose:
```
p2p = { path = "../..", version = "0.2.0", package = "tentacle",  features = ["molc"] }
```

Comment (driftluo) [reply Cargo.toml]: I think we should have no default tendency, more is user choice.

and with default, dependence must write by:
```toml
p2p = { path = "../..", version = "0.2.0", package = "tentacle", default-features = false, features = ["molc"] }
```

but if there is no default choose:
```
p2p = { path = "../..", version = "0.2.0", package = "tentacle",  features = ["molc"] }
```

Comment (doitian) [reply Cargo.toml]: Maybe we need to add such kind of code snippet to make it fail fast?

```
#[cfg(all(not(feature = "flatc"), not(feature = "molc")))]
compile_error!("Please choose a serialization format via feature. Possible choices: flatc, molc");
```

File: .cargo/config [hunk 8]
```diff
@@ -0,0 +1,2 @@
+[build]
+target-dir= "../target"
```
Comment (doitian) [inline .cargo/config]: How about the top-level crate? Will it use the target directory in the parent directory?

Comment (driftluo) [reply .cargo/config]: all build cache will store on `../target`

**Pull request merged**